Source/WebKit/UIProcess/API/glib/WebKitClipboardPermissionRequest.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2023 Igalia S.L.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "WebKitClipboardPermissionRequest.h"
+
+#include "WebKitClipboardPermissionRequestPrivate.h"
+#include "WebKitPermissionRequest.h"
+#include <wtf/glib/WTFGType.h>
+
+#if !ENABLE(2022_GLIB_API)
+typedef WebKitPermissionRequestIface WebKitPermissionRequestInterface;
+#endif
+
+/**
+ * WebKitClipboardPermissionRequest:
+ * @See_also: #WebKitPermissionRequest, #WebKitWebView
+ *
+ * A permission request for reading clipboard contents.
+ *
+ * WebKitClipboardPermissionRequest represents a request for
+ * permission to decide whether WebKit can access the clibpard to read
+ * its contents through the Async Clipboard API.
+ *
+ * When a WebKitClipboardPermissionRequest is not handled by the user,
+ * it is denied by default.
+ *
+ * Since: 2.42
+ */
+
+static void webkit_permission_request_interface_init(WebKitPermissionRequestInterface*);
+
+struct _WebKitClipboardPermissionRequestPrivate {
+    CompletionHandler<void(WebCore::DOMPasteAccessResponse)> completionHandler;
+};
+
+WEBKIT_DEFINE_FINAL_TYPE_WITH_CODE(
+    WebKitClipboardPermissionRequest, webkit_clipboard_permission_request, G_TYPE_OBJECT, GObject,
+    G_IMPLEMENT_INTERFACE(WEBKIT_TYPE_PERMISSION_REQUEST, webkit_permission_request_interface_init))
+
+static void webkitClipboardPermissionRequestAllow(WebKitPermissionRequest* request)
+{
+    ASSERT(WEBKIT_IS_CLIPBOARD_PERMISSION_REQUEST(request));
+
+    WebKitClipboardPermissionRequestPrivate* priv = WEBKIT_CLIPBOARD_PERMISSION_REQUEST(request)->priv;
+
+    if (auto completionHandler = std::exchange(priv->completionHandler, nullptr))
+        completionHandler(WebCore::DOMPasteAccessResponse::GrantedForGesture);
```

Comment:
You don't need the `std::exchange()`, the [overloaded call operator](https://github.com/WebKit/WebKit/blob/ee0988b18d7348a73ce7db08beafbaf92788a627/Source/WTF/wtf/CompletionHandler.h#L75) already does it.

```suggestion
    if (priv->completionHandler)
        priv->completionHandler(WebCore::DOMPasteAccessResponse::GrantedForGesture);
```

Reply:
Ok.

Reply:
Unfortunately the `std::exchange` was needed here. Now it will assert with "Completion handler should not be called more than once." This requires follow-up.

Reply:
But that means that it's getting called twice, signalling a bug elsewhere. Completion handlers should not be called more than once after all!

Reply:
The bug is right here. It's going to be called multiple times because you told Carlos to remove this `std::exchange` that was needed to prevent that. It's called for the first time when the application denies the permission request, then called again at least once, but potentially many times, in dispose, once for each call to dispose.

Reply:
After the first call to `CompletionHandler::operator()`, its `m_function` member will be set to `nullptr` (that's the `std::exchange` inside the operator) and from that point onwards the `operator bool()` overload ([just two lines above](https://github.com/WebKit/WebKit/blob/ee0988b18d7348a73ce7db08beafbaf92788a627/Source/WTF/wtf/CompletionHandler.h#L69)) will return `false`, which means the `if (priv->completionHandler)` check is `false` and this code will _not_ be using the operator again. The bug has to be elsewhere.

Reply:
Ohhhh... OK then, you are correct. Huh.

I didn't actually notice this crashing. Was just looking at the code. Today I learned. ;)